extendr-api/src/optional/faer.rs

```diff
@@ -0,0 +1,253 @@
+        RMatrix::new_matrix(value.nrows(), value.ncols(), |i, j| value.read(i, j))
+    }
+}
+
+impl From<Mat<f64>> for Robj {
+    fn from(value: Mat<f64>) -> Self {
+        RMatrix::<f64>::from(value).into()
+    }
+}
+
+/// Convert a `faer::Mat<f64>` into an `RMatrix<f64>` which is not NA aware.
+impl From<MatRef<'_, f64>> for RMatrix<f64> {
+    /// Convert a faer MatRef<f64> into Robj.
+    fn from(value: MatRef<'_, f64>) -> Self {
+        RMatrix::new_matrix(value.nrows(), value.ncols(), |i, j| value.read(i, j))
+    }
+}
+
+impl From<MatRef<'_, f64>> for Robj {
+    fn from(value: MatRef<'_, f64>) -> Self {
+        RMatrix::<f64>::from(value).into_robj()
+    }
+}
+
+impl From<Mat<f64>> for RMatrix<Rfloat> {
+    fn from(value: Mat<f64>) -> Self {
+        RMatrix::new_matrix(value.nrows(), value.ncols(), |i, j| value.read(i, j).into())
+    }
+}
+
+impl From<MatRef<'_, f64>> for RMatrix<Rfloat> {
+    fn from(value: MatRef<f64>) -> Self {
+        RMatrix::new_matrix(value.nrows(), value.ncols(), |i, j| {
+            Rfloat::from(value.read(i, j))
+        })
+    }
+}
+
+impl From<RMatrix<f64>> for Mat<f64> {
+    fn from(value: RMatrix<f64>) -> Self {
+        let nrow = value.nrows();
+        let ncol = value.ncols();
+        let slice = value.as_real_slice().expect("RMatrix should be doubles");
+        Mat::from_fn(nrow, ncol, |i, j| slice[i + j * nrow])
+    }
+}
+
+impl<'a> From<&'_ RMatrix<f64>> for MatRef<'a, f64> {
+    fn from(value: &RMatrix<f64>) -> Self {
+        let nrow = value.nrows();
+        let ncol = value.ncols();
+        let slice = value.as_typed_slice().expect("RMatrix should be doubles");
+        let mat_ref = faer::mat::from_column_major_slice::<f64>(slice, nrow, ncol);
+        mat_ref
+    }
+}
+
+impl TryFrom<&Robj> for Mat<f64> {
+    type Error = Error;
+
+    fn try_from(robj: &Robj) -> Result<Self> {
+        let rmat = &RMatrix::<f64>::from_robj(robj)?;
+        let nrow = rmat.nrows();
+        let ncol = rmat.ncols();
+        let slice = rmat
+            .as_real_slice()
+            .expect("RMatrix should be double values");
+        let fmat = Mat::from_fn(nrow, ncol, |i, j| slice[i + j * nrow]);
+        Ok(fmat)
+    }
+}
+
+impl<'a> TryFrom<&'_ Robj> for MatRef<'a, f64> {
+    type Error = Error;
+
+    fn try_from(robj: &Robj) -> Result<Self> {
+        if !robj.is_matrix() {
+            return Err(Error::ExpectedMatrix(robj.clone()));
+        }
+        let dim = robj.dim().expect("robj should be a matrix");
+        let nrows = (*dim).first().expect("dimension must exist").inner() as usize;
+        let ncols = (*dim).last().expect("dimension must exist").inner() as usize;
+
+        if let Some(slice) = robj.as_typed_slice() {
+            let fmat = mat::from_column_major_slice::<f64>(slice, nrows, ncols);
+            Ok(fmat)
+        } else {
+            Err(Error::ExpectedReal(robj.clone()))
+        }
+    }
+}
+
+impl FromRobj<'_> for Mat<f64> {
```

Comment:
These work now! Maybe you can add a better error message to them? I just used the generic one..